# **Autogen RoundRobin Group Agent**
- A team that runs a group chat with participants taking turns in a round-robin fashion to publish a message to all.

In [1]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.ollama import OllamaChatCompletionClient
import asyncio
from dotenv import load_dotenv
import os
load_dotenv()

True

## **Importing Model**

In [2]:
ollama_model_clint = OllamaChatCompletionClient(model="llama3.1")

## **Define 3 specific Agent**
- dsa solver
- code reviewer
- code editor

In [3]:
dsa_solver = AssistantAgent(
    name="Complex_DSA_solver",
    model_client=ollama_model_clint,
    description="A DSA solver",
    system_message="You give code in python to solve complex dsa problems. Give under 100 words"
)


code_reviewer = AssistantAgent(
    name="Code_Reviewer",
    model_client=ollama_model_clint,
    description="A code Reviewer",
    system_message="You review the code given by the complex_dsa_solver and make sure it is optimized. Given under 20 words."
)

code_editor = AssistantAgent(
    name="Code_Editor",
    model_client=ollama_model_clint,
    description="A code Editor.",
    system_message="YOu make the code easy to understand and add comments wherever required. Give under 20 words."
)

In [4]:
from autogen_agentchat.teams import RoundRobinGroupChat

team = RoundRobinGroupChat(
    participants=[dsa_solver, code_reviewer, code_editor],
    max_turns= 3
)

In [7]:
from autogen_agentchat.messages import TextMessage
async def run_team():
    task = TextMessage(content="Write a simple code in python that can add 2 number.", source="user")
    
    result = await team.run(task=task)
    print(result)
    
await run_team()

messages=[TextMessage(id='4b7731d7-325e-40b3-b411-25361a298507', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 19, 8, 24, 13, 175400, tzinfo=datetime.timezone.utc), content='Write a simple code in python that can add 2 number.', type='TextMessage'), TextMessage(id='ab06c111-da2f-4aa7-9477-4ddd77719626', source='Complex_DSA_solver', models_usage=RequestUsage(prompt_tokens=45, completion_tokens=146), metadata={}, created_at=datetime.datetime(2025, 7, 19, 8, 24, 29, 572412, tzinfo=datetime.timezone.utc), content='**Simple Addition Code in Python**\n====================================\n\n```python\ndef add_numbers(a, b):\n    """\n    Returns the sum of two numbers.\n\n    Args:\n        a (int): First number.\n        b (int): Second number.\n\n    Returns:\n        int: Sum of a and b.\n    """\n    return a + b\n\n# Example usage:\nnum1 = 5\nnum2 = 7\nresult = add_numbers(num1, num2)\nprint(f"{num1} + {num2} = {result}")\n```\n\nThis code defines 

messages=

[TextMessage(id='4b7731d7-325e-40b3-b411-25361a298507', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 19, 8, 24, 13, 175400, tzinfo=datetime.timezone.utc), content='Write a simple code in python that can add 2 number.', type='TextMessage'), 

TextMessage(id='ab06c111-da2f-4aa7-9477-4ddd77719626', source='Complex_DSA_solver', models_usage=RequestUsage(prompt_tokens=45, completion_tokens=146), metadata={}, created_at=datetime.datetime(2025, 7, 19, 8, 24, 29, 572412, tzinfo=datetime.timezone.utc), content='**Simple Addition Code in Python**\n====================================\n\n```python\ndef add_numbers(a, b):\n    """\n    Returns the sum of two numbers.\n\n    Args:\n        a (int): First number.\n        b (int): Second number.\n\n    Returns:\n        int: Sum of a and b.\n    """\n    return a + b\n\n# Example usage:\nnum1 = 5\nnum2 = 7\nresult = add_numbers(num1, num2)\nprint(f"{num1} + {num2} = {result}")\n```\n\nThis code defines a function `add_numbers` that takes two arguments and returns their sum. It then uses this function to add two example numbers and prints the result.', type='TextMessage'), 

TextMessage(id='7707a9b7-8290-4a4b-ad64-c572eba0c724', source='Code_Reviewer', models_usage=RequestUsage(prompt_tokens=197, completion_tokens=307), metadata={}, created_at=datetime.datetime(2025, 7, 19, 8, 24, 49, 480656, tzinfo=datetime.timezone.utc), content='The given code is well-structured, readable, and efficient for its purpose. However, there are minor suggestions for improvement:\n\n*   Consider adding type hints for the function parameters and return value. For example:\n    ```python\ndef add_numbers(a: int, b: int) -> int:\n```\n    This can help with code readability and make it more Pythonic.\n\n*   The docstring could be improved to provide more context about what the function does. For example:\n\n    ```python\ndef add_numbers(a: int, b: int) -> int:\n    """\n    Calculates the sum of two integers.\n\n    Args:\n        a (int): The first integer.\n        b (int): The second integer.\n\n    Returns:\n        int: The sum of a and b.\n    """\n```\n\nHere is the updated code:\n\n```python\ndef add_numbers(a: int, b: int) -> int:\n    """\n    Calculates the sum of two integers.\n\n    Args:\n        a (int): The first integer.\n        b (int): The second integer.\n\n    Returns:\n        int: The sum of a and b.\n    """\n    return a + b\n\n# Example usage:\nnum1 = 5\nnum2 = 7\nresult = add_numbers(num1, num2)\nprint(f"{num1} + {num2} = {result}")\n```\n\nOverall, the code is already quite good for its purpose. With these minor adjustments, it becomes more readable and Pythonic.', type='TextMessage'), 

TextMessage(id='dcd21179-8d2a-4d57-945f-750b34dd20b2', source='Code_Editor', models_usage=RequestUsage(prompt_tokens=499, completion_tokens=207), metadata={}, created_at=datetime.datetime(2025, 7, 19, 8, 25, 3, 883936, tzinfo=datetime.timezone.utc), content='The provided code is well-structured and follows best practices. I\'ve added comments to explain each step in detail.\n\n```python\n# Define a function called add_numbers that takes two parameters: a and b.\ndef add_numbers(a: int, b: int) -> int:\n    """\n    Calculates the sum of two integers.\n\n    Args:\n        a (int): The first integer.\n        b (int): The second integer.\n\n    Returns:\n        int: The sum of a and b.\n    """\n\n    # Return the sum of a and b by using the built-in addition operator (+).\n    return a + b\n\n# Example usage:\nnum1 = 5\nnum2 = 7\n\n# Call the add_numbers function with num1 and num2 as arguments, and store the result in the variable \'result\'.\nresult = add_numbers(num1, num2)\n\n# Print out the result of the addition using an f-string.\nprint(f"{num1} + {num2} = {result}")\n```', type='TextMessage')] stop_reason='Maximum number of turns 3 reached.'